The node management in the forest can deadlock

After the introduction of #718 a deadlock in of of the unit tests was detected by Github [CI](TestForest_WriteBufferRecoveryIsThreadSafe).

The involved threads exhibit the following stack traces:
```
panic: test timed out after 30m0s
running tests:
	TestForest_WriteBufferRecoveryIsThreadSafe (29m49s)

goroutine 11070 [running]:
testing.(*M).startAlarm.func1()
	/opt/hostedtoolcache/go/1.21.5/x64/src/testing/testing.go:2259 +0x3b9
created by time.goFunc
	/opt/hostedtoolcache/go/1.21.5/x64/src/time/sleep.go:176 +0x2d

goroutine 1 [chan receive, 29 minutes]:
testing.(*T).Run(0xc000084820, {0x72d6e5?, 0x4ebfbc?}, 0x7f7c60)
	/opt/hostedtoolcache/go/1.21.5/x64/src/testing/testing.go:1649 +0x3c8
testing.runTests.func1(0xa619a0?)
	/opt/hostedtoolcache/go/1.21.5/x64/src/testing/testing.go:20[54](https://github.com/Fantom-foundation/Carmen/actions/runs/7473023244/job/20336367454#step:11:55) +0x3e
testing.tRunner(0xc000084820, 0xc000050c48)
	/opt/hostedtoolcache/go/1.21.5/x64/src/testing/testing.go:1595 +0xff
testing.runTests(0xc00009a0a0?, {0xa4b4c0, 0x126, 0x126}, {0x415ebf?, 0xc000050d08?, 0xa610c0?})
	/opt/hostedtoolcache/go/1.21.5/x64/src/testing/testing.go:2052 +0x445
testing.(*M).Run(0xc00009a0a0)
	/opt/hostedtoolcache/go/1.21.5/x64/src/testing/testing.go:1925 +0x636
main.main()
	_testmain.go:647 +0x19c

goroutine 10849 [semacquire, 29 minutes]:
sync.runtime_Semacquire(0xc000016370?)
	/opt/hostedtoolcache/go/1.21.5/x64/src/runtime/sema.go:62 +0x25
sync.(*WaitGroup).Wait(0x1?)
	/opt/hostedtoolcache/go/1.21.5/x64/src/sync/waitgroup.go:116 +0x48
github.com/Fantom-foundation/Carmen/go/state/mpt.TestForest_WriteBufferRecoveryIsThreadSafe(0xc0001f6340)
	/home/runner/work/Carmen/Carmen/go/state/mpt/forest_test.go:640 +0x746
testing.tRunner(0xc0001f6340, 0x7f7c60)
	/opt/hostedtoolcache/go/1.21.5/x64/src/testing/testing.go:1595 +0xff
created by testing.(*T).Run in goroutine 1
	/opt/hostedtoolcache/go/1.21.5/x64/src/testing/testing.go:1648 +0x3ad

goroutine 11067 [sync.Mutex.Lock, 29 minutes]:
sync.runtime_SemacquireMutex(0xc0000d0a80?, 0x0?, 0xc00006aa50?)
	/opt/hostedtoolcache/go/1.21.5/x64/src/runtime/sema.go:77 +0x25
sync.(*Mutex).lockSlow(0xc0000d0a90)
	/opt/hostedtoolcache/go/1.21.5/x64/src/sync/mutex.go:171 +0x15d
sync.(*Mutex).Lock(...)
	/opt/hostedtoolcache/go/1.21.5/x64/src/sync/mutex.go:90
sync.(*RWMutex).Lock(0xc0001c42d0?)
	/opt/hostedtoolcache/go/1.21.5/x64/src/sync/rwmutex.go:147 +0x31
github.com/Fantom-foundation/Carmen/go/state/mpt/shared.(*Shared[...]).GetWriteHandle(...)
	/home/runner/work/Carmen/Carmen/go/state/mpt/shared/shared.go:158
github.com/Fantom-foundation/Carmen/go/state/mpt.(*Forest).getWriteAccess.func1(0xc0000d0a80)
	/home/runner/work/Carmen/Carmen/go/state/mpt/forest.go:678 +0x45
github.com/Fantom-foundation/Carmen/go/state/mpt.getAccess[...](0xc000261f50?, 0x0?, 0xc0001bdd80?, 0x7f8210, {{0x309050b04050509}})
	/home/runner/work/Carmen/Carmen/go/state/mpt/forest.go:623 +0x6e
github.com/Fantom-foundation/Carmen/go/state/mpt.(*Forest).getWriteAccess(0x5a6fa0?, 0xc0003cc400?)
	/home/runner/work/Carmen/Carmen/go/state/mpt/forest.go:673 +0x54
github.com/Fantom-foundation/Carmen/go/state/mpt.(*Forest).SetAccountInfo(0x0?, 0x0?, {0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...}, ...)
	/home/runner/work/Carmen/Carmen/go/state/mpt/forest.go:305 +0x6d
github.com/Fantom-foundation/Carmen/go/state/mpt.TestForest_WriteBufferRecoveryIsThreadSafe.func3()
	/home/runner/work/Carmen/Carmen/go/state/mpt/forest_test.go:632 +0x23d
created by github.com/Fantom-foundation/Carmen/go/state/mpt.TestForest_WriteBufferRecoveryIsThreadSafe in goroutine 10849
	/home/runner/work/Carmen/Carmen/go/state/mpt/forest_test.go:625 +0x64f

goroutine 11065 [sync.Mutex.Lock, 29 minutes]:
sync.runtime_SemacquireMutex(0xc0000d0a80?, 0x0?, 0xc00006a8f0?)
	/opt/hostedtoolcache/go/1.21.5/x64/src/runtime/sema.go:77 +0x25
sync.(*Mutex).lockSlow(0xc0000d0a90)
	/opt/hostedtoolcache/go/1.21.5/x64/src/sync/mutex.go:171 +0x15d
sync.(*Mutex).Lock(...)
	/opt/hostedtoolcache/go/1.21.5/x64/src/sync/mutex.go:90
sync.(*RWMutex).Lock(0xc0001c42d0?)
	/opt/hostedtoolcache/go/1.21.5/x64/src/sync/rwmutex.go:147 +0x31
github.com/Fantom-foundation/Carmen/go/state/mpt/shared.(*Shared[...]).GetWriteHandle(...)
	/home/runner/work/Carmen/Carmen/go/state/mpt/shared/shared.go:158
github.com/Fantom-foundation/Carmen/go/state/mpt.(*Forest).getWriteAccess.func1(0xc0000d0a80)
	/home/runner/work/Carmen/Carmen/go/state/mpt/forest.go:678 +0x45
github.com/Fantom-foundation/Carmen/go/state/mpt.getAccess[...](0xc000261f50?, 0x0?, 0xc000052d80?, 0x7f8210, {{0xf0f090b0f000f01}})
	/home/runner/work/Carmen/Carmen/go/state/mpt/forest.go:623 +0x6e
github.com/Fantom-foundation/Carmen/go/state/mpt.(*Forest).getWriteAccess(0x5a6fa0?, 0xc0003cc400?)
	/home/runner/work/Carmen/Carmen/go/state/mpt/forest.go:673 +0x54
github.com/Fantom-foundation/Carmen/go/state/mpt.(*Forest).SetAccountInfo(0x0?, 0x0?, {0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...}, ...)
	/home/runner/work/Carmen/Carmen/go/state/mpt/forest.go:305 +0x6d
github.com/Fantom-foundation/Carmen/go/state/mpt.TestForest_WriteBufferRecoveryIsThreadSafe.func3()
	/home/runner/work/Carmen/Carmen/go/state/mpt/forest_test.go:632 +0x23d
created by github.com/Fantom-foundation/Carmen/go/state/mpt.TestForest_WriteBufferRecoveryIsThreadSafe in goroutine 10849
	/home/runner/work/Carmen/Carmen/go/state/mpt/forest_test.go:625 +0x64f

goroutine 11058 [chan receive, 29 minutes]:
github.com/Fantom-foundation/Carmen/go/state/mpt.makeForest.func1()
	/home/runner/work/Carmen/Carmen/go/state/mpt/forest.go:270 +0x109
created by github.com/Fantom-foundation/Carmen/go/state/mpt.makeForest in goroutine 10849
	/home/runner/work/Carmen/Carmen/go/state/mpt/forest.go:266 +0x849

goroutine 11062 [sync.Mutex.Lock, 29 minutes]:
sync.runtime_SemacquireMutex(0xc0000d0a80?, 0x80?, 0x0?)
	/opt/hostedtoolcache/go/1.21.5/x64/src/runtime/sema.go:77 +0x25
sync.(*Mutex).lockSlow(0xc0000d0a90)
	/opt/hostedtoolcache/go/1.21.5/x64/src/sync/mutex.go:171 +0x15d
sync.(*Mutex).Lock(...)
	/opt/hostedtoolcache/go/1.21.5/x64/src/sync/mutex.go:90
sync.(*RWMutex).Lock(0xc0001c42d0?)
	/opt/hostedtoolcache/go/1.21.5/x64/src/sync/rwmutex.go:147 +0x31
github.com/Fantom-foundation/Carmen/go/state/mpt/shared.(*Shared[...]).GetWriteHandle(...)
	/home/runner/work/Carmen/Carmen/go/state/mpt/shared/shared.go:158
github.com/Fantom-foundation/Carmen/go/state/mpt.(*Forest).getWriteAccess.func1(0xc0000d0a80)
	/home/runner/work/Carmen/Carmen/go/state/mpt/forest.go:678 +0x45
github.com/Fantom-foundation/Carmen/go/state/mpt.getAccess[...](0x0?, 0xc000185d78?, 0xc000057d80?, 0x7f8210, {{0xc000185d98}})
	/home/runner/work/Carmen/Carmen/go/state/mpt/forest.go:623 +0x6e
github.com/Fantom-foundation/Carmen/go/state/mpt.(*Forest).getWriteAccess(0x1?, 0x0?)
	/home/runner/work/Carmen/Carmen/go/state/mpt/forest.go:673 +0x54
github.com/Fantom-foundation/Carmen/go/state/mpt.(*Forest).SetAccountInfo(0x5a32ba?, 0x409140?, {0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...}, ...)
	/home/runner/work/Carmen/Carmen/go/state/mpt/forest.go:305 +0x6d
github.com/Fantom-foundation/Carmen/go/state/mpt.TestForest_WriteBufferRecoveryIsThreadSafe.func3()
	/home/runner/work/Carmen/Carmen/go/state/mpt/forest_test.go:632 +0x23d
created by github.com/Fantom-foundation/Carmen/go/state/mpt.TestForest_WriteBufferRecoveryIsThreadSafe in goroutine 10849
	/home/runner/work/Carmen/Carmen/go/state/mpt/forest_test.go:625 +0x64f

goroutine 11059 [sync.Mutex.Lock, 29 minutes, locked to thread]:
sync.runtime_SemacquireMutex(0xc00010bde0?, 0x56?, 0xc00010beb8?)
	/opt/hostedtoolcache/go/1.21.5/x64/src/runtime/sema.go:77 +0x25
sync.(*Mutex).lockSlow(0xc0000d0a90)
	/opt/hostedtoolcache/go/1.21.5/x64/src/sync/mutex.go:171 +0x15d
sync.(*Mutex).Lock(...)
	/opt/hostedtoolcache/go/1.21.5/x64/src/sync/mutex.go:90
sync.(*RWMutex).Lock(0x6c8ac0?)
	/opt/hostedtoolcache/go/1.21.5/x64/src/sync/rwmutex.go:147 +0x31
github.com/Fantom-foundation/Carmen/go/state/mpt/shared.(*Shared[...]).GetWriteHandle(...)
	/home/runner/work/Carmen/Carmen/go/state/mpt/shared/shared.go:158
github.com/Fantom-foundation/Carmen/go/state/mpt.(*writeBuffer).emptyBuffer(0xc0000ba070)
	/home/runner/work/Carmen/Carmen/go/state/mpt/write_buffer.go:193 +0x2f6
github.com/Fantom-foundation/Carmen/go/state/mpt.makeWriteBuffer.func1()
	/home/runner/work/Carmen/Carmen/go/state/mpt/write_buffer.go:101 +0x111
created by github.com/Fantom-foundation/Carmen/go/state/mpt.makeWriteBuffer in goroutine 10849
	/home/runner/work/Carmen/Carmen/go/state/mpt/write_buffer.go:85 +0x1a8

goroutine 11066 [chan send, 29 minutes]:
github.com/Fantom-foundation/Carmen/go/state/mpt.(*writeBuffer).Add(0xc0000ba070, 0x3, 0xc0000d0a80)
	/home/runner/work/Carmen/Carmen/go/state/mpt/write_buffer.go:120 +0x110
github.com/Fantom-foundation/Carmen/go/state/mpt.(*Forest).addToCache(0xc0001c4200, 0xc00018b470?, 0x0?)
	/home/runner/work/Carmen/Carmen/go/state/mpt/forest.go:747 +0xfa
github.com/Fantom-foundation/Carmen/go/state/mpt.(*Forest).getSharedNode(0xc0001c4200, 0xc00018b470)
	/home/runner/work/Carmen/Carmen/go/state/mpt/forest.go:[55](https://github.com/Fantom-foundation/Carmen/actions/runs/7473023244/job/20336367454#step:11:56)9 +0x865
github.com/Fantom-foundation/Carmen/go/state/mpt.getAccess[...](0x0?, 0xc000135d78?, 0xc000135d80?, 0x7f8210, {{0xc000135d98}})
	/home/runner/work/Carmen/Carmen/go/state/mpt/forest.go:614 +0x36
github.com/Fantom-foundation/Carmen/go/state/mpt.(*Forest).getWriteAccess(0x0?, 0x0?)
	/home/runner/work/Carmen/Carmen/go/state/mpt/forest.go:673 +0x54
github.com/Fantom-foundation/Carmen/go/state/mpt.(*Forest).SetAccountInfo(0x0?, 0x0?, {0x1, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...}, ...)
	/home/runner/work/Carmen/Carmen/go/state/mpt/forest.go:305 +0x6d
github.com/Fantom-foundation/Carmen/go/state/mpt.TestForest_WriteBufferRecoveryIsThreadSafe.func3()
	/home/runner/work/Carmen/Carmen/go/state/mpt/forest_test.go:632 +0x23d
created by github.com/Fantom-foundation/Carmen/go/state/mpt.TestForest_WriteBufferRecoveryIsThreadSafe in goroutine 10849
	/home/runner/work/Carmen/Carmen/go/state/mpt/forest_test.go:625 +0x64f

goroutine 11063 [sync.Mutex.Lock, 29 minutes]:
sync.runtime_SemacquireMutex(0xc000138dc0?, 0xa5?, 0x6c8d[60](https://github.com/Fantom-foundation/Carmen/actions/runs/7473023244/job/20336367454#step:11:61)?)
	/opt/hostedtoolcache/go/1.21.5/x64/src/runtime/sema.go:77 +0x25
sync.(*Mutex).lockSlow(0xc0001c42d0)
	/opt/hostedtoolcache/go/1.21.5/x64/src/sync/mutex.go:171 +0x15d
sync.(*Mutex).Lock(...)
	/opt/hostedtoolcache/go/1.21.5/x64/src/sync/mutex.go:90
github.com/Fantom-foundation/Carmen/go/state/mpt.(*Forest).getSharedNode(0xc0001c4200, 0xc0002462a0)
	/home/runner/work/Carmen/Carmen/go/state/mpt/forest.go:553 +0xcc
github.com/Fantom-foundation/Carmen/go/state/mpt.getAccess[...](0xc0002[61](https://github.com/Fantom-foundation/Carmen/actions/runs/7473023244/job/20336367454#step:11:62)f50?, 0x0?, 0xc000139d80?, 0x7f8210, {{0x7000e00080f020a}})
	/home/runner/work/Carmen/Carmen/go/state/mpt/forest.go:[62](https://github.com/Fantom-foundation/Carmen/actions/runs/7473023244/job/20336367454#step:11:63)4 +0x85
github.com/Fantom-foundation/Carmen/go/state/mpt.(*Forest).getWriteAccess(0x5a6fa0?, 0xc0003cc400?)
	/home/runner/work/Carmen/Carmen/go/state/mpt/forest.go:673 +0x54
github.com/Fantom-foundation/Carmen/go/state/mpt.(*Forest).SetAccountInfo(0x0?, 0x0?, {0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...}, ...)
	/home/runner/work/Carmen/Carmen/go/state/mpt/forest.go:305 +0x6d
github.com/Fantom-foundation/Carmen/go/state/mpt.TestForest_WriteBufferRecoveryIsThreadSafe.func3()
	/home/runner/work/Carmen/Carmen/go/state/mpt/forest_test.go:[63](https://github.com/Fantom-foundation/Carmen/actions/runs/7473023244/job/20336367454#step:11:64)2 +0x23d
created by github.com/Fantom-foundation/Carmen/go/state/mpt.TestForest_WriteBufferRecoveryIsThreadSafe in goroutine 10849
	/home/runner/work/Carmen/Carmen/go/state/mpt/forest_test.go:625 +0x[64](https://github.com/Fantom-foundation/Carmen/actions/runs/7473023244/job/20336367454#step:11:65)f
FAIL	github.com/Fantom-foundation/Carmen/go/state/mpt	1[80](https://github.com/Fantom-foundation/Carmen/actions/runs/7473023244/job/20336367454#step:11:81)0.055s
```

--- a/go/state/mpt/forest.go
+++ b/go/state/mpt/forest.go
@@ -54,8 +54,9 @@ type Root struct {
 // the functional and non-functional properties of a forest but do not change
 // the on-disk format.
 type ForestConfig struct {
-	Mode          StorageMode // whether to perform destructive or constructive updates
-	CacheCapacity int         // the maximum number of nodes retained in memory
+	Mode                   StorageMode // whether to perform destructive or constructive updates
+	CacheCapacity          int         // the maximum number of nodes retained in memory
+	writeBufferChannelSize int         // the maximum number of elements retained in the write buffer channel
 }
 
 // Forest is a utility node managing nodes for one or more Tries.
@@ -288,7 +289,12 @@ func makeForest(
 		}
 	}()
 
-	res.writeBuffer = makeWriteBuffer(writeBufferSink{res}, 1024)
+	channelSize := forestConfig.writeBufferChannelSize
+	if channelSize <= 0 {
+		channelSize = 1024 // the default value
+	}
+
+	res.writeBuffer = makeWriteBuffer(writeBufferSink{res}, channelSize)
 	return res, nil
 }
 

--- a/go/state/mpt/forest_test.go
+++ b/go/state/mpt/forest_test.go
@@ -596,6 +596,15 @@ func testForest_WriteBufferRecoveryIsThreadSafe(t *testing.T, withConcurrentNode
 	accounts := stock.NewMockStock[uint64, AccountNode](ctrl)
 	values := stock.NewMockStock[uint64, ValueNode](ctrl)
 
+	branches.EXPECT().Flush()
+	branches.EXPECT().Close()
+	extensions.EXPECT().Flush()
+	extensions.EXPECT().Close()
+	accounts.EXPECT().Flush()
+	accounts.EXPECT().Close()
+	values.EXPECT().Flush()
+	values.EXPECT().Close()
+
 	forest, err := makeForest(
 		MptConfig{Hashing: DirectHashing},
 		t.TempDir(),
@@ -672,6 +681,10 @@ func testForest_WriteBufferRecoveryIsThreadSafe(t *testing.T, withConcurrentNode
 	}
 
 	wg.Wait()
+
+	if err := forest.Close(); err != nil {
+		t.Errorf("failed to close forest: %v", err)
+	}
 }
 
 // openFileShadowForest creates a forest instance based on a shadowed file
@@ -718,4 +731,99 @@ func openFileShadowForest(directory string, mptConfig MptConfig, forestConfig Fo
 	accounts := shadow.MakeShadowStock(accountsA, accountsB)
 	values := shadow.MakeShadowStock(valuesA, valuesB)
 	return makeForest(mptConfig, directory, branches, extensions, accounts, values, forestConfig)
+}
+
+func TestForest_NodeHandlingDoesNotDeadlock(t *testing.T) {
+	// This test used to trigger a bug leading to a deadlock as reported in
+	// https://github.com/Fantom-foundation/Carmen/issues/724
+	//
+	// It runs a stress test on the forest's node management code with a
+	// particular focus on the transitioning of nodes between a tiny cache
+	// and the write buffer. To increase the likelihood of the issue to
+	// occur, the size of the channel used to transfer nodes from the cache
+	// to the write buffer is reduced to 1.
+	//
+	// For a full description of the identified deadlock see the comments of
+	// issue https://github.com/Fantom-foundation/Carmen/issues/724
+
+	ctrl := gomock.NewController(t)
+
+	branches := stock.NewMockStock[uint64, BranchNode](ctrl)
+	extensions := stock.NewMockStock[uint64, ExtensionNode](ctrl)
+	accounts := stock.NewMockStock[uint64, AccountNode](ctrl)
+	values := stock.NewMockStock[uint64, ValueNode](ctrl)
+
+	branches.EXPECT().Flush()
+	branches.EXPECT().Close()
+	extensions.EXPECT().Flush()
+	extensions.EXPECT().Close()
+	accounts.EXPECT().Flush()
+	accounts.EXPECT().Close()
+	values.EXPECT().Flush()
+	values.EXPECT().Close()
+
+	forest, err := makeForest(
+		MptConfig{Hashing: DirectHashing},
+		t.TempDir(),
+		branches,
+		extensions,
+		accounts,
+		values,
+		ForestConfig{
+			CacheCapacity:          1,
+			writeBufferChannelSize: 1,
+		},
+	)
+	if err != nil {
+		t.Fatalf("failed to create test forest: %v", err)
+	}
+
+	var counter atomic.Uint64
+	accounts.EXPECT().New().AnyTimes().DoAndReturn(func() (uint64, error) {
+		return counter.Add(1) - 1, nil
+	})
+	accounts.EXPECT().Get(gomock.Any()).AnyTimes().DoAndReturn(func(i uint64) (AccountNode, error) {
+		return AccountNode{address: common.Address{byte(i)}, info: AccountInfo{Nonce: common.Nonce{byte(i)}}}, nil
+	})
+
+	// We need to set expectations for output operations to avoid a panic in the node writer
+	// goroutine which causes a deadlock since held locks are not properly released while resolving
+	// the panic triggered by a missing expectation.
+	accounts.EXPECT().Set(gomock.Any(), gomock.Any()).AnyTimes()
+
+	empty := NewNodeReference(EmptyId())
+	treeA, err := forest.SetAccountInfo(&empty, common.Address{0}, AccountInfo{Nonce: common.Nonce{1}})
+	if err != nil {
+		t.Fatalf("failed to create tree A: %v", err)
+	}
+
+	treeB, err := forest.SetAccountInfo(&empty, common.Address{1}, AccountInfo{Nonce: common.Nonce{2}})
+	if err != nil {
+		t.Fatalf("failed to create tree B: %v", err)
+	}
+
+	const N = 10
+	var wg sync.WaitGroup
+	wg.Add(N)
+	for i := 0; i < N; i++ {
+		go func() {
+			defer wg.Done()
+			for i := 0; i < 1000; i++ {
+				tree := treeA
+				if i%2 > 0 {
+					tree = treeB
+				}
+				_, err := forest.SetAccountInfo(&tree, common.Address{byte(i % 2)}, AccountInfo{Nonce: common.Nonce{byte(i%200 + 2)}})
+				if err != nil {
+					t.Errorf("failed to update trie: %v", err)
+				}
+			}
+		}()
+	}
+
+	wg.Wait()
+
+	if err := forest.Close(); err != nil {
+		t.Errorf("failed to close forest: %v", err)
+	}
 }

--- a/go/state/mpt/write_buffer.go
+++ b/go/state/mpt/write_buffer.go
@@ -51,17 +51,18 @@ func MakeWriteBuffer(sink NodeSink) WriteBuffer {
 // ----------------------------------------------------------------------------
 
 type writeBuffer struct {
-	sink        NodeSink
-	capacity    int
-	buffer      map[NodeId]*shared.Shared[Node]
-	bufferMutex sync.Mutex
-	ids         chan NodeId
-	idsMutex    sync.Mutex
-	idsClosed   bool
-	flushDone   <-chan struct{}
-	done        <-chan struct{}
-	errs        []error
-	errsMutex   sync.Mutex
+	sink                    NodeSink
+	capacity                int
+	counter                 int
+	buffer                  map[NodeId]*shared.Shared[Node]
+	bufferMutex             sync.Mutex
+	emptyBufferSignal       chan bool // true if an explicit flush is triggered, false for an implicit
+	emptyBufferSignalMutex  sync.Mutex
+	emptyBufferSignalClosed bool
+	flushDone               <-chan struct{}
+	done                    <-chan struct{}
+	errs                    []error
+	errsMutex               sync.Mutex
 }
 
 func makeWriteBuffer(sink NodeSink, capacity int) WriteBuffer {
@@ -69,37 +70,30 @@ func makeWriteBuffer(sink NodeSink, capacity int) WriteBuffer {
 		capacity = 1
 	}
 
-	ids := make(chan NodeId, capacity)
+	emptyBufferSignal := make(chan bool, 1)
 	flushDone := make(chan struct{})
 	done := make(chan struct{})
 
 	res := &writeBuffer{
-		sink:      sink,
-		capacity:  capacity,
-		buffer:    make(map[NodeId]*shared.Shared[Node], 2*capacity),
-		ids:       ids,
-		flushDone: flushDone,
-		done:      done,
+		sink:              sink,
+		capacity:          capacity,
+		buffer:            make(map[NodeId]*shared.Shared[Node], 2*capacity),
+		emptyBufferSignal: emptyBufferSignal,
+		flushDone:         flushDone,
+		done:              done,
 	}
 
+	// A background goroutine flushing nodes to the sink and handling
+	// synchronization tasks.
 	go func() {
 		runtime.LockOSThread()
 		defer runtime.UnlockOSThread()
 		defer close(done)
 		defer close(flushDone)
-		counter := 0
-		for id := range ids {
-			// Check if this is a token signaling a flush request.
-			if id.IsEmpty() {
-				res.emptyBuffer()
+		for flush := range emptyBufferSignal {
+			res.emptyBuffer()
+			if flush {
 				flushDone <- struct{}{}
-				continue
-			}
-			// Otherwise, perform a flush every time the capacity is reached.
-			counter++
-			if counter >= capacity {
-				res.emptyBuffer()
-				counter = 0
 			}
 		}
 	}()
@@ -115,11 +109,20 @@ func (b *writeBuffer) Add(id NodeId, node *shared.Shared[Node]) {
 	b.bufferMutex.Lock()
 	b.buffer[id] = node
 	b.bufferMutex.Unlock()
-	b.idsMutex.Lock()
-	if !b.idsClosed {
-		b.ids <- id
+	b.emptyBufferSignalMutex.Lock()
+	b.counter++
+	if b.counter > b.capacity && !b.emptyBufferSignalClosed {
+		// The option to ignore a full signal channel here is important
+		// to prevent a potential deadlock. See
+		// https://github.com/Fantom-foundation/Carmen/issues/724
+		// for more details.
+		select {
+		case b.emptyBufferSignal <- false: /* ok, a new clear-buffer operation is scheduled */
+		default: /* also fine, an operation to clear the buffer is already pending */
+		}
+		b.counter = 0
 	}
-	b.idsMutex.Unlock()
+	b.emptyBufferSignalMutex.Unlock()
 }
 
 func (b *writeBuffer) contains(id NodeId) bool {
@@ -141,11 +144,11 @@ func (b *writeBuffer) Cancel(id NodeId) (*shared.Shared[Node], bool) {
 }
 
 func (b *writeBuffer) Flush() error {
-	b.idsMutex.Lock()
-	if !b.idsClosed {
-		b.ids <- EmptyId()
+	b.emptyBufferSignalMutex.Lock()
+	if !b.emptyBufferSignalClosed {
+		b.emptyBufferSignal <- true
 	}
-	b.idsMutex.Unlock()
+	b.emptyBufferSignalMutex.Unlock()
 	<-b.flushDone // finishes either due to flush signal or being closed
 	b.errsMutex.Lock()
 	defer b.errsMutex.Unlock()
@@ -153,12 +156,12 @@ func (b *writeBuffer) Flush() error {
 }
 
 func (b *writeBuffer) Close() error {
-	b.idsMutex.Lock()
-	if !b.idsClosed {
-		close(b.ids)
-		b.idsClosed = true
+	b.emptyBufferSignalMutex.Lock()
+	if !b.emptyBufferSignalClosed {
+		close(b.emptyBufferSignal)
+		b.emptyBufferSignalClosed = true
 	}
-	b.idsMutex.Unlock()
+	b.emptyBufferSignalMutex.Unlock()
 	<-b.done // finishes once all elements are written
 	b.errsMutex.Lock()
 	defer b.errsMutex.Unlock()
@@ -205,7 +208,7 @@ func (b *writeBuffer) emptyBuffer() {
 		delete(b.buffer, id)
 		b.bufferMutex.Unlock()
 
-		// Only release read access of the node after it has been removed from
+		// Only release access of the node after it has been removed from
 		// the buffer such that subsequent updates are not lost.
 		handle.Release()
 	}

